components/plan-build/bin/hab-plan-build.sh

```diff
@@ -570,7 +570,7 @@ _assemble_runtime_path() {
       data="$(cat "$dep_prefix/PATH")"
       data="$(trim "$data")"
       while read -r entry; do
-        paths=($(_return_or_append_to_set "$entry" "${paths[@]}"))
+        paths+=$(_return_or_append_to_set "$entry" "${paths[@]}")
```

Comment:
Is this going to do what I think? I've been staring at this for too long. @baumanj

Reply:
i think i might need `paths=$(_return_or_append_to_set "$entry" "${paths[@]}")`?

Reply:
First an aside: A super powerful bash debugging tool is `declare -p` (intuitive, I know 🙄), which will dump a variable in a format appropriate to put in code. It will show the distinction between a string and array that would appear identical it just `echo`ed:
```
$ foo_array=(a b c)
$ foo_string="a b c"
$ foo_array_string=("a b c")

$ echo $foo_array
a
$ echo "$foo_string"
a b c
$ echo "$foo_array_string"
a b c

$ echo "${foo_array[@]}"
a b c
$ echo "${foo_string[@]}"
a b c
$ echo "${foo_array_string[@]}"
a b c

$ declare -p foo_array
declare -a foo_array=([0]="a" [1]="b" [2]="c")
$ declare -p foo_string
declare -- foo_string="a b c"
$ declare -p foo_array_string
declare -a foo_array_string=([0]="a b c")
```

Back to this code. This is subtle and `_return_or_append_to_set` is *really* easy to misuse. I _don't_ think you want the current change:
```bash
paths=$(_return_or_append_to_set "$entry" "${paths[@]}")
```
To explain why: start a new bash shell and paste in `_return_or_append_to_set` and `_array_contains` (which it depends on) so we can play with them. Then, let's define some test data:
```bash
$ paths=(a b c)
$ entry=d
```
Now let's see what happens with the code as it was:
```bash
$ old_code_paths=($(_return_or_append_to_set "$entry" "${paths[@]}"))
$ declare -p old_code_paths
declare -a old_code_paths=([0]="a" [1]="b" [2]="c" [3]="d")
```
and with the current change:
```bash
$ new_code_paths=$(_return_or_append_to_set "$entry" "${paths[@]}")
$ declare -p new_code_paths
declare -- new_code_paths="a b c d"
```
We wanted to keep `paths` an array, but now it's a string. The problem is that there's no such thing as "returning an array" in bash. In fact, there's no such thing as return values aside from exit status. What `_return_or_append_to_set` does in this example is print out "a b c d", but when evaluated in the context of an array `(…)`, that gets split into 4 words, and we get a 4 element array. It's exactly equivalent to:
```bash
$ foo_array=($(echo "a b c d"))
$ declare -p foo_array
declare -a foo_array=([0]="a" [1]="b" [2]="c" [3]="d")
```
The fact that `_return_or_append_to_set` contains the line
```bash
echo "${appended_set[@]}"
```
Is pretty misleading. It could be
```bash
echo "${appended_set[*]}"
```
and it wouldn't change anything. It's the responsibility of the caller to re-interpret the output this function does as an array.

So, the reason we're changing the old code in the first place is to fix [this clippy lint (2207)](https://github.com/koalaman/shellcheck/wiki/SC2207), right? It warns against code like:
```bash
array=( $(mycommand) )
```
Which our existing code definitely is. It suggests three alternatives, depending on the nature of `mycommand` (i.e., `_return_or_append_to_set`):
1. If it outputs multiple lines, each of which should be an element
1. If it outputs a line with multiple words (separated by spaces, other delimiters can be chosen with IFS), each of which should be an element
1. If the output should be a single element

We're case (2) (and we can use the default `IFS` value, since that's what `_return_or_append_to_set` does), so I think what we want is:
```bash
$ read -r -a paths_with_lint_fix <<< "$(_return_or_append_to_set "$entry" "${paths[@]}")"
$ declare -p paths_with_lint_fix
declare -a paths_with_lint_fix=([0]="a" [1]="b" [2]="c" [3]="d")
```
or in the actual code:
```bash
        read -r -a paths <<< "$(_return_or_append_to_set "$entry" "${paths[@]}")"
```
Which `shellcheck` smiles approvingly at.

Yes, it's kind of ugly, but I'd say that's largely because `_return_or_append_to_set`'s interface is awkward. I'd prefer to see it redesigned to accept the name of the array that it should mutate rather than requiring the caller to recreate the array from the string it outputs.